trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/library/ILibraryModel.java

```diff
@@ -27,16 +27,63 @@
 package de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.library;
 
 import java.util.Collection;
+import java.util.List;
 
-import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.library.LibraryModelHandler.IFunctionModelHandler;
+import org.eclipse.cdt.core.dom.ast.IASTFunctionCallExpression;
+
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.IDispatcher;
 import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.container.c.ICType;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.ExpressionResult;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.Result;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
 
 /**
  * An interface to abstract the model of libraries (mostly libraries from the C standard) in Boogie.
  *
  * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
  */
 public interface ILibraryModel {
+	/**
+	 * An interface to represent the model of a library function.
+	 *
+	 * @author Daniel Dietsch (dietsch@informatik.uni-freiburg.de)
+	 */
+	@FunctionalInterface
+	interface IFunctionModelHandler {
+		/**
+		 * Translates a library function.
+		 *
+		 * @param main
+		 *            A dispatcher.
+		 * @param node
+		 *            A node for the function call.
+		 * @param loc
+		 *            A location.
+		 * @param methodName
+		 *            The name of the called function.
+		 * @return The model of the call to a library function.
+		 */
+		Result handleFunction(final IDispatcher main, final IASTFunctionCallExpression node, final ILocation loc,
+				String methodName);
+	}
+
+	/**
+	 * An interface to represent the model of a constant macro.
+	 *
+	 * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
+	 */
+	@FunctionalInterface
+	interface IConstantModelHandler {
+		/**
+		 * Models the constant as an {@link ExpressionResult}.
+		 *
+		 * @param loc
+		 *            A location.
+		 * @return An {@link ExpressionResult} as a model of the constant.
+		 */
+		ExpressionResult handleConstant(ILocation loc);
```

Comment:
Is there a specific reason why the return type here is more restrictive than in `IFunctionModelHandler`? In principle, it makes sense to me that for a constant, an expression is returned. But what are the functions for which this is not the case? Could any of these scenarios also apply to constants?

Reply:
What do you mean with "more restrictive"? Because `ExpressionResult` is returned here instead of just `Result`? An arbitrary `Result` does not really make sense for constants, as an `LRValue` should be returned, which is only possible for an `ExpressionResult`. On the either hand, to model `void` function it might be perfectly fine to return a `SkipResult`.

And it does also not work to return just a  `LRValue`, as the handling of  `__FUNCTION__` in (`GccBuiltinLibraryModel`) requires an aux-var and maybe we need the flexibility to use additional Boogie statements in the future.